### ISSUE TITLE
Run on iOS12.4 or older will crash on app start

### ISSUE THREAD

**Issue by haowei-jj** [RenderHeads/UnityPlugin-AVProVideo#810]
We found that the plugin run on iOS12.4 or older will crash on start. The log is as follow:

14:49:23.273360+0800  log  Error loading /var/containers/Bundle/Application/7C700117-F3C0-4FE9-8A24-19A76F8EE2CC/log.app/Frameworks/UnityFramework.framework/UnityFramework:  dlopen(/var/containers/Bundle/Application/7C700117-F3C0-4FE9-8A24-19A76F8EE2CC/log.app/Frameworks/UnityFramework.framework/UnityFramework, 265): Library not loaded: @rpath/libswiftCore.dylib
  Referenced from: /private/var/containers/Bundle/Application/7C700117-F3C0-4FE9-8A24-19A76F8EE2CC/log.app/Frameworks/AVProVideo.framework/AVProVideo
  Reason: image not found

 - Unity version: 2019.4.10
 - AVPro Video version: 2.0.8
 - Operating system version: iOS 12.2
 - Device model: iphone 7 plus
 - Video specs (resolution, frame-rate, codec, file size): 

There is also a weird thing that if we run the app with xCode, everything is ok, but if we download the app from appstore or testflight it will crash.

**Comment by kahnivore** [comment]
Hi @haowei-jj 

Could you please refer to [this issue response](https://github.com/RenderHeads/UnityPlugin-AVProVideo/issues/680#issuecomment-801897379) and let us know if it helps? 

Thanks

**Comment by haowei-jj** [comment]
Hi
Setting the AVProVideo.framework to Embed & Sign not solve my issue!



***@***.***
 
From: kahnivore
Date: 2021-06-03 20:28
To: RenderHeads/UnityPlugin-AVProVideo
CC: haowei-jj; Mention
Subject: Re: [RenderHeads/UnityPlugin-AVProVideo] Run on iOS12.4 or older will crash on app start (#810)
Hi @haowei-jj
Could you please refer to this issue response and let us know if it helps?
Thanks

**Comment by kahnivore** [comment]
I was referring to setting "Always Embed Swift Standard Libraries" in the Build Settings.

![swift](https://user-images.githubusercontent.com/44468274/120648614-21fab780-c47c-11eb-9ddc-12fe7db5cf0e.png)

**Comment by haowei-jj** [comment]
Hi,
In Unity-iPhone and UnityFramework all have this option. Should I all set them to YES?

Thanks！


***@***.***
 
From: kahnivore
Date: 2021-06-03 20:58
To: RenderHeads/UnityPlugin-AVProVideo
CC: haowei-jj; Mention
Subject: Re: [RenderHeads/UnityPlugin-AVProVideo] Run on iOS12.4 or older will crash on app start (#810)
I was referring to setting "Always Embed Swift Standard Libraries" in the Build Settings.

**Comment by AstroViking** [comment]
Hi @haowei-jj @kahnivore  

We are struggling with the exact same issue currently.

* Unity version: 2020.3.9f1
* AVPro Video versions tried: 2.1.0 and 2.1.3 
* Operating system version: iOS 12.0
* Device model: iPhone X

We switched from the watermarked Demo Version to the paid Core version on May 5th. However we hadn't noticed it immediately because we hadn't tested it on iOS < 13  until much later.

* Demo Version 2.0.10 was working fine.
Version 2.1.0 and 2.1.3 are crashing on launch.

* We can confirm this: 
> There is also a weird thing that if we run the app with xCode, everything is ok, but if we download the app from appstore or testflight it will crash.

* `Always Embed Swift Standard Libraries` is set, but doesn't make any difference (wether we set it for the Unity-iPhone target, the UnityFramework target or even both).

* We have tried nearly everything from that post to fix it, but to no avail: https://stackoverflow.com/questions/26024100/dyld-library-not-loaded-rpath-libswiftcore-dylib

* The only thing that at least made a difference was adding a dummy swift file to force the inclusion of the swift libraries according to: https://stackoverflow.com/a/61935640
However, instead of fixing the issue, it only led to a new one with packing/signing the Xcode archive into an IPA file for ad-hoc distribution with the following error message:

```
021-06-03 13:02:30 +0000  ipatool JSON: {
    alerts =     (
                {
            code = 3762;
            description = "Configuration issue: platform AppleTVSimulator.platform doesn't have any non-simulator SDKs; ignoring it";
            info =             {
            };
            level = WARN;
        },
                {
            code = 3762;
            description = "Configuration issue: platform WatchSimulator.platform doesn't have any non-simulator SDKs; ignoring it";
            info =             {
            };
            level = WARN;
        },
                {
            code = 3762;
            description = "Configuration issue: platform iPhoneSimulator.platform doesn't have any non-simulator SDKs; ignoring it";
            info =             {
            };
            level = WARN;
        },
                {
            code = 431;
            description = "Failed to resolve rpath for AVProVideo: Could not resolve @executable_path for @executable_path/Frameworks from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo armv7 -> @rpath/libswiftAVFoundation.dylib: Could not resolve @rpath in @rpath/libswiftAVFoundation.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo armv7 -> @rpath/libswiftCoreAudio.dylib: Could not resolve @rpath in @rpath/libswiftCoreAudio.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo armv7 -> @rpath/libswiftCoreMedia.dylib: Could not resolve @rpath in @rpath/libswiftCoreMedia.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo armv7 -> @rpath/libswiftsimd.dylib: Could not resolve @rpath in @rpath/libswiftsimd.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 431;
            description = "Failed to resolve rpath for AVProVideo: Could not resolve @executable_path for @executable_path/Frameworks from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo arm64 -> @rpath/libswiftAVFoundation.dylib: Could not resolve @rpath in @rpath/libswiftAVFoundation.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo arm64 -> @rpath/libswiftCoreAudio.dylib: Could not resolve @rpath in @rpath/libswiftCoreAudio.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo arm64 -> @rpath/libswiftCoreMedia.dylib: Could not resolve @rpath in @rpath/libswiftCoreMedia.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 455;
            description = "Failed to resolve linkage dependency AVProVideo arm64 -> @rpath/libswiftsimd.dylib: Could not resolve @rpath in @rpath/libswiftsimd.dylib from AVProVideo";
            info =             {
            };
            level = WARN;
        },
                {
            code = 431;
            description = "Failed to resolve rpath for UnityFramework: Could not resolve @executable_path for @executable_path/Frameworks from UnityFramework";
            info =             {
            };
            level = WARN;
        },
                {
            code = 2161;
            description = "Could not determine platform for arm64 in /var/folders/v4/kcmft70s0bxggk3btttx0nkm0000gp/T/ipatool20210603-93857-fq62p5/thinned-out/arm64/Payload/Ambia.app/Ambia.dSYM/Contents/Resources/DWARF/Ambia";
            info =             {
            };
            level = WARN;
        },
                {
            code = 2161;
            description = "Could not determine platform for arm64 in /var/folders/v4/kcmft70s0bxggk3btttx0nkm0000gp/T/ipatool20210603-93857-fq62p5/thinned-out/arm64/Payload/Ambia.app/Frameworks/libswiftCore.dylib.dSYM/Contents/Resources/DWARF/libswiftCore.dylib";
            info =             {
            };
            level = WARN;
        }
    );
}
```

**Comment by haowei-jj** [comment]
Hi,
I found a weird thing that the develop version has all the .dylib about Swift, but in the archive there is nothing. Maybe this will cause the libswiftCore.dylib not loaded. Could you have any idea about this?

Thanks!


***@***.***
 
From: kahnivore
Date: 2021-06-03 20:58
To: RenderHeads/UnityPlugin-AVProVideo
CC: haowei-jj; Mention
Subject: Re: [RenderHeads/UnityPlugin-AVProVideo] Run on iOS12.4 or older will crash on app start (#810)
I was referring to setting "Always Embed Swift Standard Libraries" in the Build Settings.

**Comment by kahnivore** [comment]
@haowei-jj @AstroViking thanks for reporting this and for the details. 

Our Mac guy will be back on Monday so will get an answer for you then.

**Comment by MorrisRH** [comment]
The problem is that the execute bits have been cleared on the plugin binary when importing from the asset store. This prevents the Xcode archive process from being able to deduce that swift is in fact being used.

We're currently investigating why this has started happening. The fix for now is to manually set the execute bits via the terminal as follows:

```
cd <path to your Unity project>
chmod a+x Assets/AVProVideo/Runtime/Plugins/iOS/AVProVideo.framework/AVProVideo
```

You still need to set "```Always Embed Swift Standard Libraries```" to "```YES```" on the Unity-iPhone target.

**Comment by haowei-jj** [comment]
OK，I'll try this. Thanks!



***@***.***
 
From: Morris Butler
Date: 2021-06-07 22:17
To: RenderHeads/UnityPlugin-AVProVideo
CC: haowei-jj; Mention
Subject: Re: [RenderHeads/UnityPlugin-AVProVideo] Run on iOS12.4 or older will crash on app start (#810)
The problem is that the execute bits have been cleared on the plugin binary when importing from the asset store. This prevents the Xcode archive process from being able to deduce that swift is in fact being used.
We're currently investigating why this has started happening. The fix for now is to manually set the execute bits via the terminal as follows:
cd <path to your Unity project>
chmod a+x Assets/AVProVideo/Runtime/Plugins/iOS/AVProVideo.framework/AVProVideo

You still need to set "Always Embed Swift Standard Libraries" to "YES" on the Unity-iPhone target.

**Comment by MorrisRH** [comment]
I've added some checks to the PostProcessBuild_iOS script to make sure the execute bits have been set on the plugin binary and to also automatically set ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES when the target OS version is less than 12.2. This will available in the next release, version 2.1.4.

**Comment by AstroViking** [comment]
I can confirm that this was indeed the issue and it is now working on our end.

Thank you for the quick support! :)

**Comment by MorrisRH** [comment]
Unfortunately the fix doesn't work for the first build after importing the plugin unitypackage, it does work for all subsequent builds though. I'm looking for a way to improve this...

**Comment by MorrisRH** [comment]
A new post-process script is here: [PostProcessBuild_iOS.cs.zip](https://github.com/RenderHeads/UnityPlugin-AVProVideo/files/6662614/PostProcessBuild_iOS.cs.zip)

Just replace ```Assets/AVProVideo/Editor/Scripts/PostProcessBuild_iOS.cs``` with it.

It will also be available in the next release (2.1.5)
